include/bout/paralleltransform.hxx

```diff
@@ -93,7 +102,13 @@ public:
 class ShiftedMetric : public ParallelTransform {
 public:
   ShiftedMetric() = delete;
-  ShiftedMetric(Mesh &mesh);
+  ShiftedMetric(Mesh &m) : mesh(m) {}
+
+  /// initialize after constructor is finished, so that
+  /// localmesh->getCoordinateSystem() can call
+  /// ParallelTransform::getCoordinateSystem() for fields belonging to the
+  /// ParallelTransform
+  void initialize();
```

Comment:
This sort of two stage setup has advantages for a number of things, but it is an important design decision so I think there would need to be quite a bit of input on this from @bendudson, @ZedThree  and others.

Reply:
@d7919 agreed. I did consider calling `setCoordinateSystem` explicitly for all the fields in each `ParallelTransform`, but it got quite messy. It could make more sense to put off this kind of redesign, e.g. if we will make `ParallelTransform`s belong to `Coordinates`, it would make sense to design the initialization of both together.

For what I need at the moment, I think could get away with reducing the amount of checking: I only really want the checks in `to/fromFieldAligned`, and it could be OK in the short term to allow some fields to float around with `coordinate_system = COORDINATE_SYSTEM::None`, especially since they would only be in the `ParallelTransforms` and are probably only accessed by index rather than with `Field3D` operations. I'll have a think if there's a sensible, simpler initial version to start with, so we can give longer consideration to something like the implementation here, which is a significant re-design.

Reply:
I've moved the split initialization into #1463. This PR still seems to work (at least passes the tests) without it. The split initialization is, I think, necessary for *all* `Field3D`s to have their `coordinate_system` set correctly, but I think the only ones that don't get one set are in `ShiftedMetric` or `FCI` and only get accessed via indices, where no check is (or can really be) performed.

I think something like it was going to be more necessary when I was trying to get `Field2D` to also have a `coordinate_system`: then all the `Field2D`s in `Coordinates` need to get set properly (which #1463 actually doesn't help with in the end...) because they multiply `Field3D`s in the differential operators, etc., so if they had `coordinate_system = COORDINATE_SYSTEM::None` then checks would fail, or some horrible hacks would be necessary, like preventing copying the `coordinate_system` from a `Field*D` which has `coordinate_system = COORDINATE_SYSTEM::None`.

Reply:
I would much prefer to avoid two-phase initialisation if possible, as it increases the surface area for potential bugs.

Would moving the parallel transform from `Mesh` into `Coordinates` help? I think that's probably where it really belongs anyway. 
Another thing that might help would be to use another type instead of `Field*`s in `Coordinates`/`ParallelTransform` that aren't "on" coordinates (though they'd still need to be "on" meshes).

Reply:
I can't actually remember now exactly what made me think two-phase initialization was necessary. I've added a few commits trying to sort this a different way, adding constructors which take an explicit CoordinateSystem for the ParallelTransform constructors to use. Also attempting to add coordinate_system to Field2D. Still needs testing, probably tweaking a bit too, but I've got to go now. I'll look over it again later, hopefully this week.